src/core/drive/head_snapshot.ts

```diff
@@ -1,71 +1,37 @@
 import { Snapshot } from "../snapshot"
 
-type ElementDetailMap = { [outerHTML: string]: ElementDetails }
-
-type ElementDetails = {
-  type?: ElementType
-  tracked: boolean
-  elements: Element[]
-}
-
-type ElementType = "script" | "stylesheet"
-
 export class HeadSnapshot extends Snapshot<HTMLHeadElement> {
-  readonly detailsByOuterHTML = this.children
-    .filter((element) => !elementIsNoscript(element))
-    .map((element) => elementWithoutNonce(element))
-    .reduce((result, element) => {
-      const { outerHTML } = element
-      const details: ElementDetails =
-        outerHTML in result
-          ? result[outerHTML]
-          : {
-              type: elementType(element),
-              tracked: elementIsTracked(element),
-              elements: [],
-            }
-      return {
-        ...result,
-        [outerHTML]: {
-          ...details,
-          elements: [...details.elements, element],
-        },
-      }
-    }, {} as ElementDetailMap)
+  elements: { [key: string]: Array<Element> } = {}
+  stylesheetElements: Array<Element> = []
+  trackedElemements: Array<Element> = []
 
-  get trackedElementSignature(): string {
-    return Object.keys(this.detailsByOuterHTML)
-      .filter((outerHTML) => this.detailsByOuterHTML[outerHTML].tracked)
-      .join("")
+  constructor(element: HTMLHeadElement) {
+    super(element)
+    this.parseDetailsByOuterHTML()
   }
 
-  getScriptElementsNotInSnapshot(snapshot: HeadSnapshot) {
-    return this.getElementsMatchingTypeNotInSnapshot<HTMLScriptElement>("script", snapshot)
+  parseDetailsByOuterHTML() {
+    for (let element of this.children) {
+      element = elementWithoutNonce(element)
+      if (elementIsStylesheet(element)) this.stylesheetElements.push(element)
+      else if (Object.prototype.hasOwnProperty.call(this.elements, element.localName))
```

Comment:
A separate "miss" conditional that uses `Object.prototype.hasOwnProperty` combined with `call` and a bound `this` feels more complicated than access-or-assignment. Have you considered combining both `else` branches?

```typescript
if (elementIsStylesheet(element)) {
  this.stylesheetElements.push(element)
} else {
  const collection = this.elements || []
  collection.push(element)
  this.elements[element.localName] = collection
}
```

Reply:
Ooh - that's nice.